Added CometLogger

dsblank opened the pull request:
# Adds CometLogger to spaCy

## Description

Like the existing WandbLogger, this adds the CometLogger.

### Types of change

This adds the CometLogger to the existing loggers, and updates the documentation to list the logger.

I used alphabetical order where possible for existing lists, but if you wish to alter that ordering, that is no problem.

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

svlandeg commented on website/docs/api/top-level.md:
```diff
@@ -488,6 +491,47 @@ logging the results.
+also details on the Python and OS environment, output, and system usage
+metrics. You can manage running experiments, see `diffs` with other
+experiments, and share analyses through custom visualizations with
+your team at [Comet.ML](https://www.comet.ml/) . Of course, you can
+control exactly which items are logged, or not, through configuration
```
```suggestion
your team at [Comet.ML](https://www.comet.ml/). You can
control which items are logged through configuration
```

svlandeg commented on website/docs/api/top-level.md:
```diff
@@ -488,6 +491,47 @@ logging the results.
+experiments, and share analyses through custom visualizations with
+your team at [Comet.ML](https://www.comet.ml/) . Of course, you can
+control exactly which items are logged, or not, through configuration
+settings. To create your Comet.ML API key, visit:
+[www.comet.ml/docs/python-sdk/spacy](https://www.comet.ml/docs/python-sdk/spacy/)
```
```suggestion
settings. To create your Comet.ML API key, visit:
[www.comet.ml/docs/python-sdk/spacy](https://www.comet.ml/docs/python-sdk/spacy/).
```

svlandeg commented on website/docs/api/top-level.md:
```diff
@@ -488,6 +491,47 @@ logging the results.
+| `project_name`         | The name of the project in your Comet.ML workspace. The project will be created automatically if it doesn't exist yet. ~~str~~ |
+| `remove_config_values` | A list of values to remove from the config before it is logged to Comet.ML (default: empty). ~~List[str]~~                              |
+
+<Project id="integrations/comet">
+
+To quickly see how easy it is to track your spaCy training runs in
+Comet.ML try our project template. It trains on the IMDB Movie Review
+Dataset and includes a simple config with the built-in `CometLogger`.
+
+</Project>
```
This needs to be removed - there isn't actually a project called `integrations/comet`, so this will create a dead link. I don't think we would want to have an individual project per type of logger, but f we start having more of these, perhaps we'd need to have a project that shows off all the integrations for experiment tracking though.

dsblank replied:
I was working on that project. But ok, I'll remove it. I guess I should also remove wandb's then as well.

dsblank replied:
Done!

svlandeg replied:
Please don't remove any of the current content - we've worked hard to get these example projects in place and to align the documentation to them. This PR should focus on getting the new logger implemented and properly documented - referring to existing projects and functionality in the docs only. If it gets merged & released, we can discuss how to potentially update the relevant example projects in a next stage.

svlandeg replied:
We can't publish a project around new functionality until the relevant new spaCy version is released, anyway, so it's not feasible to push all this in one PR.

dsblank replied:
How about if I add the `integrations/comet` project first, and then I can refer to it? I'm just trying to keep these two logger examples aligned.

svlandeg replied:
That doesn't work from a technical point of view. The functionality needs to be in spaCy first, in a published release, until a spacy project can refer to it.

dsblank replied:
Ok, I think I have addressed all of the concerns. Thanks for the feedback!

svlandeg commented on website/docs/usage/projects.md:
```diff
@@ -43,11 +43,12 @@ and experiments, iterate on demos and prototypes and ship your models into
 
 <Grid narrow cols={3}>
+<Integration title="Comet" logo="comet" url="#comet">Manage your experiments and models</Integration>
 <Integration title="DVC" logo="dvc" url="#dvc">Manage and version your data</Integration>
-<Integration title="Prodigy" logo="prodigy" url="#prodigy">Create labelled training data</Integration>
-<Integration title="Streamlit" logo="streamlit" url="#streamlit">Visualize and demo your pipelines</Integration>
 <Integration title="FastAPI" logo="fastapi" url="#fastapi">Serve your models and host APIs</Integration>
+<Integration title="Prodigy" logo="prodigy" url="#prodigy">Create labelled training data</Integration>
 <Integration title="Ray" logo="ray" url="#ray">Distributed and parallel training</Integration>
+<Integration title="Streamlit" logo="streamlit" url="#streamlit">Visualize and demo your pipelines</Integration>
 <Integration title="Weights &amp; Biases" logo="wandb" url="#wandb">Track your experiments and results</Integration>
 </Grid>
```
There isn't a Comet project, so the entry here should be removed. Let's just keep the original ordering as well.

dsblank replied:
Ok, I'll remove the wandb Project as well.

svlandeg replied:
No - please don't - cf my comment above.

dsblank replied:
Fixed!

svlandeg commented on website/docs/usage/projects.md:
```diff
@@ -227,6 +228,37 @@ pipelines.
 | `workflows`     | A dictionary of workflow names, mapped to a list of command names, to execute in order. Workflows can be run with the [`project run`](/api/cli#project-run) command.                                                                                                                                                                                                                        ...
 | `commands`      | A list of named commands. A command can define an optional help message (shown in the CLI when the user adds `--help`) and the `script`, a list of commands to run. The `deps` and `outputs` let you define the created file the command depends on and produces, respectively. This lets spaCy determine whether a command needs to be re-run because its dependencies or outputs changed. ...
 | `spacy_version` | Optional spaCy version range like `>=3.0.0,<3.1.0` that the project is compatible with. If it's loaded with an incompatible version, an error is raised when the project is loaded.                                                                                                                                                                                                         ...
+### Comet {#comet} <IntegrationLogo name="comet" width={175} height="auto" align="right" />
```
The `Data assets` subsection should definitely come first, this part of the docs is still describing how the projects work. 

Basically this subsection should be moved to the very end, right after "Weights & Biases".

dsblank replied:
Done!

svlandeg commented on website/docs/usage/projects.md:
```diff
@@ -227,6 +228,37 @@ pipelines.
 | `spacy_version` | Optional spaCy version range like `>=3.0.0,<3.1.0` that the project is compatible with. If it's loaded with an incompatible version, an error is raised when the project is loaded.                                                                                                                                                                                                         ...
+### Comet {#comet} <IntegrationLogo name="comet" width={175} height="auto" align="right" />
+
+[Comet.ML](https://www.comet.ml/), like spaCy itself, is out of this
+world. Comet enables teams to track, explain, and optimize spaCy experiments
```
As a Wikipedia adminstrator would say - this is not quite [NPOV](https://en.wikipedia.org/wiki/Wikipedia:Neutral_point_of_view) ;-) Let's just remove the first sentence here.

dsblank replied:
It was just making a reference to their space-oriented naming. I'll remove.

dsblank replied:
Done!

svlandeg commented on website/docs/usage/projects.md:
```diff
@@ -227,6 +228,37 @@ pipelines.
+![Screenshot: Track Metrics](../images/comet1.png)
+![Screenshot: Confusion Matrix and Panels](../images/comet2.png)
+
+<Project id="integrations/comet">
+
+Get started with tracking your spaCy training runs in Comet.ML using our
+project template. It trains on the IMDB Movie Review Dataset and includes a
+simple config with the built-in `CometLogger`, as well as a custom example of
+creating variants of the config for a simple hyperparameter grid search and
+logging the results.
+
+</Project>
```
Again this reference to a non-existing project should be removed

dsblank replied:
Again, I'll add the project.

dsblank replied:
We'll add the project after the CometLogger is released.

svlandeg commented on spacy/training/loggers.py:
```diff
@@ -137,3 +137,77 @@ def finalize() -> None:
+            print("You need to configure Comet, and run appropriately: see: https://comet.ml/docs/python-sdk/spacy/")
+
+        # Get the config, removing items if necessary:
+        config = nlp.config.interpolate()
+        config_dot = util.dict_to_dot(config)
+        for field in remove_config_values:
+            del config_dot[field]
+        config = util.dot_to_dict(config_dot)
```
I think this code is identical to what the W&B logger is using - so perhaps it makes sense to move this to a helper function, so that all loggers easily have access to that functionality and we avoid duplication?

dsblank replied:
Moved to utils

svlandeg commented on spacy/training/loggers.py:
```diff
@@ -137,3 +137,77 @@ def finalize() -> None:
+        # Get methods for step console processing:
+        console_log_step, console_finalize = console(nlp, stdout, stderr)
+
+        def flatten_dictionary(prefix, dictionary, results):
+            """
+            Recursively construct key/value pair.
+            """
+            for key, value in dictionary.items():
+                if isinstance(value, dict):
+                    if prefix:
+                        flatten_dictionary(prefix + "/" + key, value, results)
+                    else:
+                        flatten_dictionary(key, value, results)
+                else:
+                    if prefix:
+                        results[prefix + "/" + key] = value
+                    else:
+                        results[key] = value
```
Again it feels like this should probably be a helper/utility function

dsblank replied:
Moved to utils

svlandeg reviewed (changes requested):
Thanks a lot for this contribution! As Comet is quite a popular tool, it makes sense to provide out-of-the-box integration. It looks like the integration is also pretty light-weight and really matches the logging API well. 

I've added some detailed comments below and will mark this as "request changes" for now.

dsblank commented on spacy/training/loggers.py:
```diff
@@ -137,3 +137,77 @@ def finalize() -> None:
+    try:
+        import comet_ml
+    except ImportError:
+        print("You need to install `comet_ml` to use the CometLogger: see: https://comet.ml/docs/python-sdk/spacy/")
```
Oh, that is new from 4 days ago. Updated!

svlandeg replied:
Ah, yep, you're right: https://github.com/explosion/spaCy/pull/7223

dsblank commented on spacy/training/loggers.py:
```diff
@@ -137,3 +137,77 @@ def finalize() -> None:
+            experiment = comet_ml.Experiment(project_name=project_name)
+        except Exception:
+            experiment = None
+            print("You need to configure Comet, and run appropriately: see: https://comet.ml/docs/python-sdk/spacy/")
```
Done!

dsblank commented on website/docs/api/top-level.md:
```diff
@@ -384,11 +384,14 @@ finished. To log each training step, a
+
+1. a logger that prints results to the console in tabular format (the default)
+2. a logger that sends the results to a [Comet.ML](https://www.comet.ml/)
+3. a logger that sends the results to a [Weights & Biases](https://www.wandb.com/) dashboard
```
Done!

dsblank commented on website/docs/api/top-level.md:
```diff
@@ -384,11 +384,14 @@ finished. To log each training step, a
+3. a logger that sends the results to a [Weights & Biases](https://www.wandb.com/) dashboard
+
+In addition to using one of the built-in loggers listed here, you
+can also [implement your own](/usage/training#custom-logging).
```
Done!

dsblank commented:
Should be ready for a final review. Thanks for feedback!

svlandeg commented on website/docs/usage/projects.md:
```diff
@@ -1004,7 +1005,7 @@ your results.
 ![Screenshot: Parameter importance using config values](../images/wandb2.jpg 'Parameter importance using config values')
 
-<Project id="integrations/wandb">
+</Project>
```
```suggestion
<Project id="integrations/wandb">
```

honnibal commented:
@dsblank I'm coming to this PR late, but I want to state forcefully that I'm very unhappy with the way you have engaged here.

Here's a review of what I've seen here:

* You reordered the content so that CometML would appear first
* You added a dead link to a project that wasn't created yet.
* When the dead link was pointed out to you, you _deleted content that mentioned the W&B integration_.
* You've included marketing language in the documentation that Sofie has had to police carefully
* You've insisted that we'll "add a project after this is in place", and _marked the issue as resolved_, even though you've not done what was clearly explained to you.
* You'e marked all the other issues as resolved yourself, making me reopen these conversations to look at them.

My inclination is to simply close this PR. It is not our problem to ensure some sort of parity in integration or prominence between all the different commercial tools. The fact that we have a W&B integration and a project demonstrating that integration does not necessitate that we have a CometML integration. We will not work with you if you come to the project with the sort of attitude I've seen in this PR.

dsblank commented:
@honnibal Sorry that you feel this way. I've been a supporter of open source software all my life, and do feel like we might have gotten off to on the wrong foot. 

I've tried to adapt to your all of your feedback (and feedback I have gotten from others as well). I apologize over any friction I have caused. I, too, found the interaction much more harsh that I have found on other projects. That was, no doubt, because of my misunderstandings of the process.

> You reordered the content so that CometML would appear first

I mentioned this in the PR description, and I am fine with whatever ordering you prefer.

> You added a dead link to a project that wasn't created yet.

As noted, I had already started on the project PR. We would like to make a project and contribute to spaCy to make it easy for others to use and discover. The project is a great idea, and I hope that you continue to provide such a service.

> When the dead link was pointed out to you, you deleted content that mentioned the W&B integration.

I was only trying to keep these two contributions on par. I replaced those immediately. Of course I knew from the start that it was a "dead project" but didn't understand your strict policy. That was my mistake. 

> You've included marketing language in the documentation that Sofie has had to police carefully

The marketing language I removed was "Comet.ML, like spaCy, is out of this world."  I immediately removed that with the feedback. I am very sorry that this was seen as marketing language. I did not mean for it to be seen that way.

> You've insisted that we'll "add a project after this is in place", and marked the issue as resolved, even though you've not done what was clearly explained to you.

I think you may be misinformed about the timing of the responses. I removed the project when it was explained what the issue was. There may have been some miscommunication. (I saw in email some github comments that did not show up in the PR until much later).

> You'e marked all the other issues as resolved yourself, making me reopen these conversations to look at them.

Sorry, I understood issues to be resolved when they were resolved to everyone's understanding. I only resolved them when I explicitly followed the specific directions, doing exactly what was required. I would not close an issue unless I thought that everyone agreed. Perhaps you have different preferences here. If I misunderstood the work flow, then I am truly sorry.

> My inclination is to simply close this PR. It is not our problem to ensure some sort of parity in integration or prominence between all the different commercial tools. The fact that we have a W&B integration and a project demonstrating that integration does not necessitate that we have a CometML integration. We will not work with you if you come to the project with the sort of attitude I've seen in this PR.

I'm new to this community. I had to make many decisions in how I added new content, and I obviously raised some red flags with you. Rest assured, that we would love to cooperate further, and we have some other suggestions for moving forward. The good news is that I think this PR was a valuable lesson for me, and hopefully provides infrastructure for the next contributor.

svlandeg commented:
> I think you may be misinformed about the timing of the responses. I removed the project when it was explained what the issue was. There may have been some miscommunication. (I saw in email some github comments that did not show up in the PR until much later).
> (...) 
> I only resolved them when I explicitly followed the specific directions, doing exactly what was required. I would not close an issue unless I thought that everyone agreed. Perhaps you have different preferences here. If I misunderstood the work flow, then I am truly sorry.

As you are implying that I misinformed Matt on this matter, I feel the need to respond.

First of all, I did not talk to Matt about the timing of the responses at all. That is also unnecessary - he can easily draw his own conclusions from the Github history.

Second, you definitely did resolve issues that were not agreed upon. You decided to remove the W&B project when I asked you to remove a link to a different, non-existing one, you went ahead and actually deleted existing content, and immediately marked the issue as resolved. By the time [I was able to respond](https://github.com/explosion/spaCy/pull/7244#discussion_r585786483), I had to first unresolve the issue. And also after that, you kept resolving issues before I had a chance to have a look, even though by then it had become clear that there was some disagreement. 

Finally, I would like to point out that I reviewed this PR in great detail to help you improve it and to make it suitable for inclusion in spaCy. There was no miscommunication - you tried pushing a link to a project that doesn't (and can't) exist yet, and when I asked you to remove it, you started deleting other content. This attitude certainly does not come off as constructive, and I fully agree with Matt that collaboration like this is not desirable. I will leave the final decision on this PR with my colleagues.

dsblank commented:
> As you are implying that I misinformed Matt on this matter,

I did not mean to imply that! I only meant that the order that *I* received the github messages may not be reflected in the order that they are in the record. I think that had misunderstood how you wanted to proceed. But I think it is exactly how you wanted it in the end. If not, I will correct the PR.

I very much appreciate your feedback on the PR.

dsblank commented:
> There was no miscommunication

There was, and I take full responsibility for that. I only ever meant to do exactly what was required.

dsblank commented:
In order to proceed in a productive way, would it be ok if I have a colleague take over from here and handle any remaining issues?

honnibal commented:
> I've tried to adapt to your all of your feedback (and feedback I have gotten from others as well). I apologize over any friction I have caused. I, too, found the interaction much more harsh that I have found on other projects. That was, no doubt, because of my misunderstandings of the process.

Come on. Please see my two screenshots attached. In the first one it is _absolutely clear_ that you should remove the content you added. I find it beyond belief that you would respond by _removing other content as well_ and then mark the issue as resolved so that the change is hidden. When Sofie objected, you then readded the original content as well.

Then see the next screenshot. You are told once more, no, remove that content. You say no and mark the issue resolved. The issue is _not_ resolved, the content is still there! Do not now act all affronted, as though you've been in good faith this whole time and it's us who have been unreasonable.
<img width="607" alt="Screenshot 2021-03-06 at 12 33 16" src="https://user-images.githubusercontent.com/8059750/110190713-c9198400-7e78-11eb-9f6d-e110341c54e6.png">
<img width="859" alt="Screenshot 2021-03-06 at 12 31 17" src="https://user-images.githubusercontent.com/8059750/110190727-d767a000-7e78-11eb-9d94-b563c3fb5413.png">

dsblank commented:
I now understand why there was such a negative response to the interaction. Looking at those screen shots, I don't blame you for any feeling that you have about my responses, and for closing this. 

I will just say that the timing of the responses is a bit mixed from the order I was receiving and responding in real time. 

In any event, I wish your project well, and wish you success!
